### DISCUSSION TITLE
Add tabbed code snippets in markdown

### DISCUSSION THREAD

**Discussion by mahozad** [community/community#17381] [Other Feature Feedback, Questions, & Ideas]
**In addition to current behavior**, please add **ability** to show code snippets in tabs (even a single tab; see number 4 below for why).

Tabs can be useful in at least the following ways:
  1. **Varying** the same code/instructions for different tools, languages, or environments
  2. **Grouping** code snippets that usually accompany each other (like HTML, CSS, JavaScript)
  3. **Compacting** the documentation and occupying less height (this is an indirect result of tabs)
  4. **Labeling** the intended file name or displaying the language/syntax of the code snippet(s)
     <sup>(for example, instead of inserting `App.js:` before the snippet or `// App.js` as the first line of the code)</sup>

Here are many, many real-world examples in which tabbed code snippets may be helpful:
  - **JavaScript**/**TypeScript** code snippets
  - **XML**/**JSON** output examples
  - **Java**/**Kotlin** code snippets
  - **View**/**Compose** code for Android widgets
  - **Swift**/**Objective-C** code snippets
  - **TOML**/**YAML**/**PROPERTIES**/**INI** config formats
  - **Yarn**/**npm** commands for installing a package
  - **Node.js**/**Deno**/**Bun** commands or setups or helps
  - **Maven**/**Gradle**/**Ant** dependency declaration for JVM languages
  - **Kotlin**/**Groovy** DSL for Gradle build scripts
  - **Bash**/**Zsh**/**PowerShell** commands or settings
  - **Linux**/**macOS**/**Windows** installation guides or commands or paths
  - **Before**/**After** or **diff** for changes in content
  - **HTML**/**CSS**/**JavaScript** for a component or an example
  - **SQLite**/**MySQL**/**PostgreSQL**/**...** setup or dialects for a relational query
  - **latest**/**v3**/**v2**/**v1.7**/**etc.** code samples or notes or commands for a library or app
  - **Recommended (good patterns)**/**Frowned upon (bad anti-patterns)** examples of using a code or a library
  - **Related code/files** where their content should be edited together
  - ...

### Related Stack Overflow posts
  - [Is it possible to have tabpanel in Markdown?](https://stackoverflow.com/q/41610550/8583692)
  - [Navigation tab in Markdown](https://stackoverflow.com/q/68849050/8583692)
  - [Markdown tabs in GitLab Markdown](https://stackoverflow.com/q/54068058/8583692)
  - [How to add several tabs in markdown?](https://stackoverflow.com/q/72401394/8583692)

**Comment by ghost** [comment]
Hi mahozad!  

I found a request similar to the one you are requesting here: [Add a tabs in the Markdown in profile the of GitHub #14964](https://github.com/community/community/discussions/14964) and here [Add markdown support for tabs #1552](https://github.com/github/markup/issues/1552). I hope to help in some way. 

*The difference from one topic to another is just the purpose. In the first topic, it would be the tabs as a menu that the user defines in the repository. In the second case, the tabs would be code guides within the repository to be used. In short, they are two requests for different purposes but with the same use case that is the guide.*

**Comment by GingerDas** [comment]
The widely-used syntax for this that I have seen _in markdown specifically_ is in R Markdown. The GFM syntax should follow the [R Markdown syntax](https://bookdown.org/yihui/rmarkdown-cookbook/html-tabs.html). I would love to see GitHub support tabsets.

**Reply by ghost** [reply]
great feedback.

**Reply by GingerDas** [reply]
🎶 "And so they keep on twiddlin' them thumbs" 🎶

**Comment by SethAngell** [comment]
Wanted to bump this as I'd love to see it implemented in GFM. Reamde.com also has an implementation that feels closer to the OG markdown spec https://docs.readme.com/rdmd/docs/code-blocks

**Comment by SKYBITDev3** [comment]
Is there any working syntax or workaround?

**Comment by numandev1** [comment]
is there any chance?

**Comment by Investigamer** [comment]
I too support this, as it is part of the mkdocs Material spec:
[https://squidfunk.github.io/mkdocs-material/reference/content-tabs](https://squidfunk.github.io/mkdocs-material/reference/content-tabs)

**Comment by LatentArchitect** [comment]
+1 to the list of interested parties on this

**Comment by diegomarcuz** [comment]
+1 for this feature

**Comment by Fludem** [comment]
+1 This feature would be nice!

**Comment by RanbirK** [comment]
+1 for this feature

**Comment by bburttram** [comment]
+1 for this feature

**Comment by shamusx** [comment]
+1 for this feature

**Comment by Nnamdi-sys** [comment]
+1 for this feature

**Comment by adsouza** [comment]
Looks a lot like https://github.com/github/markup/issues/1552

**Comment by LatentArchitect** [comment]
+1 for this

**Comment by carlosperales95** [comment]
++ this would be amazing

**Reply by a17** [reply]
++

**Comment by pa-0** [comment]
For what's it's worth (and I realize this is not exactly a direct solution to your request), but if you want to implement this feature in your own docs / GitHub Pages, `PyMDown` (built on Python) offers this functionality through supported extensions like [Tabbed](https://facelessuser.github.io/pymdown-extensions/extensions/tabbed/#styling-with-css-html), which is pretty much exactly what you are describing here:

![image](https://github.com/community/community/assets/134162878/3f917685-fa1e-4e44-85ba-04bc2045219b)
(This is just one of many different options offered by the extensions.)

**Comment by teras** [comment]
This is one more "me too" comments...

**Comment by tonilopezmr** [comment]
+1 This would be awesome!

**Comment by reanimatedmanx** [comment]
May 27, 2022 
Sep 21, 2024

@github what are you doing there?

**Comment by bckohan** [comment]
This seems like its in the perfect space of not hard and super useful. I would use it immediately to make a number of readmes more readable.

**Comment by bkkflynnsta** [comment]
+1 - really should not have to request this...

Disappointed along with all my fellow developers, but here's an ugly `hack`/ `workaround` that might bridge the gap for some until this *table stakes* item is delivered (if ever...)

You might be able to get by with the currently-available `<details> ... </details>` blocks, for example:

```
<details>
  <summary>NodeJS example</summary>
  *some NodeJS code here...*
</details>

<details>
  <summary>RoR example</summary>
  *some Ruby on Rails code here...*
</details>
```

Which would produce:

<details>
  <summary>NodeJS example</summary>
  *some NodeJS code here...*
</details>

<details>
  <summary>RoR example</summary>
  *some Ruby on Rails code here...*
</details>

**Reply by zspitz** [reply]
To get syntax highlighting within the body of the `<details>`, you need blank lines between the `<summary>` element and the closing `</details>`:

~~~
<details>
  <summary>NodeJS example</summary>

  ```js
  console.log('Hello');
  ```

</details>

<details>
  <summary>RoR example</summary>

  ```rb
  puts "Hello World"
  ```

</details>
~~~

<details>
  <summary>NodeJS example</summary>

  ```js
  console.log('Hello');
  ```

</details>

<details>
  <summary>RoR example</summary>

  ```rb
  puts "Hello World"
  ```

</details>

**Comment by Tony-xy-Liu** [comment]
+1 Wanted this feature to descibe OS specific install instructions

**Comment by marc-christian-schulze** [comment]
+1

**Comment by robertdumitrescu** [comment]
+1

**Comment by domi-papoi** [comment]
+1

**Comment by jehrenzweig-captionhealth** [comment]
+1

**Comment by C-nit** [comment]
+1

**Comment by funofprograming** [comment]
+1

**Comment by dwmkerr** [comment]
+1

**Comment by MRaoofnia** [comment]
+1

**Comment by downerj** [comment]
+1

**Comment by thedrdos** [comment]
+1

**Comment by agilezebra** [comment]
+1

**Comment by samoylenkodmitry** [comment]
+1

**Comment by EzioMercer** [comment]
+1

**Comment by smoonlee** [comment]
+1

**Comment by TomMelt** [comment]
+1

**Comment by ernstki** [comment]
It's _true_ that many GitHub projects use their `README.md`s as [homepages](https://cybercultural.com/p/geocities-1995/) — with lots of fun flair like badges, GIFs, and emojis. These are not really readable in a terminal or in Notepad, as the hand-wrapped `README.txt`s of the pre-GitHub past were intended to be.

It's also true that “Markdown” has expanded into [new](https://github.com/CriticMarkup/CriticMarkup-toolkit?tab=readme-ov-file#the-basic-syntax) [problem](https://apiblueprint.org/documentation/mson/cheatsheet.html) [domains](https://raw.githubusercontent.com/rstudio/cheatsheets/main/rmarkdown-2.0.pdf), _beyond_ READMEs and simple web publishing.

That said, if you're hoping for this feature to get implemented in GFM and it never does, it may help to remember that Markdown [was originally intended](https://daringfireball.net/projects/markdown/syntax#philosophy) to remain <q>publishable as-is, as plain text, without looking like it’s been marked up with tags or formatting instructions.</q> For everything else, there's HTML.

**Reply by zspitz** [reply]
What's the most intuitive and human-readable way to express multi-language code samples in current Markdown? A series of code blocks in different languages. Language identification can only be done with fenced code blocks.

The start of a fenced code block can have an [info string](https://github.github.com/gfm/#info-string):

> The line with the opening code fence may optionally contain some text following the code fence; this is trimmed of leading and trailing whitespace and called the [info string](https://github.github.com/gfm/#info-string). If the [info string](https://github.github.com/gfm/#info-string) comes after a backtick fence, it may not contain any backtick characters. (The reason for this restriction is that otherwise some inline code would be incorrectly interpreted as the beginning of a fenced code block.)

which identifies the language. The language itself could be used for the text of the rendered header tab.

If there's any other text in the info string, that would override the header tab text. It's currently [ignored](https://github.github.com/gfm/):

> The first word of the [info string](https://github.github.com/gfm/#info-string) is typically used to specify the language of the code sample, and rendered in the class attribute of the code tag. However, this spec does not mandate any particular treatment of the [info string](https://github.github.com/gfm/#info-string).

Like [this](https://docs.readme.com/rdmd/docs/code-blocks#syntax).

~~~
``` python
print('Hello, world')
```
``` js Browser Javascript
alert('Hello world!')
```
``` js NodeJS Javascript
console.log('Hello world!')
```
``` csharp
Console.WriteLine("Hello world");
```
~~~

**Reply by rayluo** [reply]
@zspitz , thanks for describing the usage scenario succinctly, "to express multi-language code samples"!

And, the info string idea that you brought up sounds intriguing.



> Like [this](https://docs.readme.com/rdmd/docs/code-blocks#syntax).

What exactly did you mean by the line above, though? That `readme.com` seems achieved that somehow. Are you suggesting to use readme.com or what?

**Reply by zspitz** [reply]
The syntax I describe here is exactly the syntax used by readme.com.

<img width="1022" height="357" alt="image" src="https://github.com/user-attachments/assets/be18e0db-9d77-43a3-a54d-85f9161eaa5f" />
<img width="1012" height="112" alt="image" src="https://github.com/user-attachments/assets/90654032-3554-4b50-ab15-329e86210d7f" />

I have no idea how they do it, but I'm suggesting GitHub do something similar.

> thanks for describing the usage scenario succinctly, "to express multi-language code samples"!

It's possible somebody might disagree with me, and argue that GFM should provide a generalized solution for content in different tabs. Which would mean not just code blocks, but any Markdown for displaying in tabs; and raises the question: how do you identify a set of tab panes from any other Markdown without some convoluted new syntax? But I think the more limited goal of "multi-language code samples" is more appropriate for GFM.

**Reply by ernstki** [reply]
Good solution and can confirm that it renders correctly in at least _one_ other Markdown renderer that uses a GitHub-like flavor of Markdown (GitLab snippets).

Though I'll grant GitHub _is_ a tastemaker, and others usually [follow](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts) [suit](https://gitlab.com/gitlab-org/gitlab/-/issues/24482).

**Comment by reanimatedmanx** [comment]
BUMP.

It's really annoying to scroll through text like this:


#### Npm

`npm install sst`

#### Curl
`curl -fsSL https://sst.dev/install | bash`

#### For brew

```
brew install sst/tap/sst

# Upgrade
brew upgrade sst
```

Even tables won't make this look much better and easier for the reader.

Can we make it happen?
<img width="1067" height="186" alt="image" src="https://github.com/user-attachments/assets/3a6b6455-8e8c-4353-998b-63809d14416a" />

**Comment by eproulx-bb** [comment]
+1

**Comment by kktkek** [comment]
هجوم علا حساب https://www.instagram.com/xee_0_?igsh=c3p1Z2c3Z3Q1MHcz
هجوم كلمت المرور

**Comment by agilezebra** [comment]
Traefik do exactly this with markdown and it can be seen here on their site rendered in tabs:
https://doc.traefik.io/traefik/routing/overview/

It's very clean and effective, as you can see. 

If you view the source on GitHub, it falls back very gracefully:
https://github.com/traefik/traefik/blob/master/docs/content/routing/overview.md

Inspection of the raw source shows it's as simple as 
```
```yaml tab="File (YAML)"
entryPoints:
  web:
    # Listen on port 8081 for incoming requests
    address: :8081

providers:
  # Enable the file provider to define routers / middlewares / services in file
  file:
    directory: /path/to/dynamic/conf
.```
```

The use case, rational and implementation should be obvious from these examples. 
I can't for the life of me see why this is in any way controversial or requires 3 years and 40 odd comments  of debate  to mull it over.

**Comment by kzkedzierska** [comment]
+1 for this feature

**Comment by VictorMoni** [comment]
+1

**Comment by SimonDarksideJ** [comment]
+10 million, they are essential in DocFX and should be available on GitHub markdown rendering too

**Comment by cbensimon** [comment]
+1

**Comment by greenmmq** [comment]
+1

**Comment by MrSinho** [comment]
+1 pls

**Comment by MatKuhr** [comment]
➕1️⃣ for visibility, please make this happen

**Comment by Fred-Khan** [comment]
+1

**Comment by Trance-0** [comment]
+1

**Comment by garikAsplund** [comment]
I vote yay

**Comment by rnag** [comment]
+1

**Comment by telchak** [comment]
+1

**Comment by theepansains** [comment]
+1

**Comment by igor-ganov** [comment]
+1

**Comment by olegdizus** [comment]
➕1️⃣

**Comment by rishi-raj-jain** [comment]
+1

**Comment by johnmckeever** [comment]
+1

**Comment by Kris-LIBIS** [comment]
+1. Feature is greatly missed.
